core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -573,6 +578,94 @@ public Journaled getDelegate() {
     return mJournaledGroup;
   }
 
+  @Override
+  public JournalContext createJournalContext() throws UnavailableException {
+    JournalContext context = super.createJournalContext();
+    if (!Configuration.getBoolean(
+        PropertyKey.MASTER_FILE_SYSTEM_MERGE_INODE_JOURNALS
+    )) {
+      return context;
+    }
+    return new FileSystemMergeJournalContext(
+        context, DefaultFileSystemMaster::mergeFileSystemJournals
+    );
+  }
+
+  /**
+   * Merge inode entry with subsequent update inode and update inode file/directory entries.
+   * This is similar to {@link InodeSyncStream#mergeCreateComplete(List)} but it also
+   * merges inode directory entries additionally
+   *
+   * @param entries list of journal entries
+   * @return a list of compacted journal entries
+   */
+  public static List<Journal.JournalEntry> mergeFileSystemJournals(
```

Comment:
I wonder if it is cleaner to merge the journal entires as they are being appended instead of at the end at flush? Also maybe this function can be in a different file.

Reply:
yeah good idea. will change